presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoinsUtils.java

```diff
@@ -0,0 +1,183 @@
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.PriorityQueue;
+import java.util.Set;
+
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.PARTITIONED;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+public class ReorderJoinsUtils
+{
+    private ReorderJoinsUtils() {}
+
+    /**
+     * This method generates all the ways of dividing  totalNodes into two sets
+     * each containing at least one node.  It will generate one set for each
+     * possible partitioning.  The other partition is implied in the absent values.
+     * In order not to generate the inverse of any set, we always include the 0th
+     * node in our sets.
+     *
+     * @param totalNodes
+     * @return A set of sets each of which defines a partitioning of totalNodes
+     */
+    public static Set<Set<Integer>> generatePartitions(int totalNodes)
+    {
+        return generatePartitions(ImmutableSet.of(0), totalNodes, 1);
+    }
+
+    private static Set<Set<Integer>> generatePartitions(Set<Integer> existingSet, int totalNodes, int index)
+    {
+        ImmutableSet.Builder<Set<Integer>> sets = ImmutableSet.builder();
+        if (index < totalNodes && existingSet.size() < totalNodes - 1) {
+            ImmutableSet.Builder<Integer> setWithAddedNode = ImmutableSet.builder();
+            setWithAddedNode.addAll(existingSet);
+            setWithAddedNode.add(index);
+            sets.addAll(generatePartitions(setWithAddedNode.build(), totalNodes, index + 1));
+            sets.addAll(generatePartitions(existingSet, totalNodes, index + 1));
+        }
+        else {
+            sets.add(existingSet);
+        }
+        return sets.build();
+    }
+
+    public static JoinNode createJoinAccordingToPartitioning(JoinGraphNode joinGraph, Set<Integer> leftSources, PlanNodeIdAllocator idAllocator)
+    {
+        List<PlanNode> sources = joinGraph.getSources();
+        ImmutableList.Builder<PlanNode> leftNodes = ImmutableList.builder();
+        List<Symbol> leftSymbols = new ArrayList<>();
+        ImmutableList.Builder<PlanNode> rightNodes = ImmutableList.builder();
+        List<Symbol> rightSymbols = new ArrayList<>();
+        for (int i = 0; i < sources.size(); i++) {
+            PlanNode source = sources.get(i);
+            if (leftSources.contains(i)) {
+                leftNodes.add(source);
+                leftSymbols.addAll(source.getOutputSymbols());
+            }
+            else {
+                rightNodes.add(source);
+                rightSymbols.addAll(source.getOutputSymbols());
+            }
+        }
+
+        ImmutableList.Builder<JoinNode.EquiJoinClause> leftCriteria = ImmutableList.builder();
```

Comment:
There seem to be quite a lot of repeated code for `EquiJoinClausea` (criteria) and for `Expressions` (filters).
Maybe it would be worthy to merge criteria and filters together for sake of computation and use just `Expression`s. And then for final result expand the ones that can be converted to EJCs.